#### Задание2. Проанализируйте набор данных «Авто-мили», чтобы выяснить, насколько переменные похожи друг на друга.
Загрузите набор данных auto-mpg из ~ / наборов данных / auto-mpg / Проверить нулевую гипотезу, отклонять или не отклонять нулевую гипотезу.

Гипотезы:
Independent variable: origin

Dependent variable: mpg

$ H_o: $ Существенных различий в милях среди транспортных средств различного происхождения не существует.

$ H_1: $ Существенная разница существует в милях среди транспортных средств различного происхождения.

Independent variable: origin

Independent variable: horsepower

Dependent variable: mpg

$ H_o: $ Существенных различий в милях среди транспортных средств различного происхождения не существует.

$ H_1: $ Существенная разница существует в милях среди транспортных средств различного происхождения.



In [96]:
import pandas as pd
import numpy as np

In [97]:
df = pd.read_csv('auto-mpg.csv')

In [98]:
df.tail(14)

mpg  cylinders  displacement horsepower  weight  acceleration  \
384  32.0          4          91.0         67    1965          15.7   
385  38.0          4          91.0         67    1995          16.2   
386  25.0          6         181.0        110    2945          16.4   
387  38.0          6         262.0         85    3015          17.0   
388  26.0          4         156.0         92    2585          14.5   
389  22.0          6         232.0        112    2835          14.7   
390  32.0          4         144.0         96    2665          13.9   
391  36.0          4         135.0         84    2370          13.0   
392  27.0          4         151.0         90    2950          17.3   
393  27.0          4         140.0         86    2790          15.6   
394  44.0          4          97.0         52    2130          24.6   
395  32.0          4         135.0         84    2295          11.6   
396  28.0          4         120.0         79    2625          18.6   
397  31.0          4         119.0         82    2720          19.4   

     model year  origin                           car name  
384          82       3                 honda civic (auto)  
385          82       3                      datsun 310 gx  
386          82       1              buick century limited  
387          82       1  oldsmobile cutlass ciera (diesel)  
388          82       1         chrysler lebaron medallion  
389          82       1                     ford granada l  
390          82       3                   toyota celica gt  
391          82       1                  dodge charger 2.2  
392          82       1                   chevrolet camaro  
393          82       1                    ford mustang gl  
394          82       2                          vw pickup  
395          82       1                      dodge rampage  
396          82       1                        ford ranger  
397          82       1                         chevy s-10

Гипотезы: Independent variable: origin

Dependent variable: mpg

In [99]:
df.origin.min()

1

In [100]:
df.groupby('origin').count()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
origin                                                                   
1       249        249           249         249     249           249   
2        70         70            70          70      70            70   
3        79         79            79          79      79            79   

        model year  car name  
origin                        
1              249       249  
2               70        70  
3               79        79

In [101]:
n1 = df.groupby('origin').count()['mpg'].values[0]
n2 = df.groupby('origin').count()['mpg'].values[1]
n3 = df.groupby('origin').count()['mpg'].values[2]

In [102]:
n = n1 + n2 + n3
n

398

In [103]:
k = 3

Origin = 1

In [104]:
y1 = df.groupby('origin').mpg.apply(list).values[0]

Origin = 2

In [105]:
y2 = df.groupby('origin').mpg.apply(list).values[1]

Origin = 3

In [106]:
y3 = df.groupby('origin').mpg.apply(list).values[2]

Проведем однофакторный дисперсионный анализ. Сначала найдем средний mpg:

In [107]:
y1_mean = np.mean(y1)
print(y1_mean)

20.083534136546184


In [108]:
y2_mean = np.mean(y2)
print(y2_mean)

27.89142857142857


In [109]:
y3_mean = np.mean(y3)
print(y3_mean)

30.450632911392404


Видно, что средние росты разнятся. Установим, что это отличие статистически значимо. Для этого сначала соберем

все значения роста в один массив:

In [110]:
y_all = np.concatenate([y1, y2, y3])
y_all

array([18. , 15. , 18. , 16. , 17. , 15. , 14. , 14. , 14. , 15. , 15. ,
       14. , 15. , 14. , 22. , 18. , 21. , 21. , 10. , 10. , 11. ,  9. ,
       28. , 25. , 19. , 16. , 17. , 19. , 18. , 14. , 14. , 14. , 14. ,
       12. , 13. , 13. , 18. , 22. , 19. , 18. , 23. , 26. , 25. , 20. ,
       21. , 13. , 14. , 15. , 14. , 17. , 11. , 13. , 12. , 13. , 15. ,
       13. , 13. , 14. , 22. , 28. , 13. , 14. , 13. , 14. , 15. , 12. ,
       13. , 13. , 14. , 13. , 12. , 13. , 18. , 16. , 18. , 18. , 23. ,
       11. , 12. , 13. , 12. , 18. , 21. , 19. , 21. , 15. , 16. , 15. ,
       11. , 20. , 21. , 19. , 15. , 26. , 25. , 16. , 16. , 18. , 16. ,
       13. , 14. , 14. , 14. , 28. , 19. , 18. , 15. , 15. , 16. , 15. ,
       16. , 14. , 17. , 16. , 15. , 18. , 21. , 20. , 13. , 23. , 20. ,
       23. , 18. , 19. , 25. , 26. , 17.5, 16. , 15.5, 14.5, 22. , 22. ,
       24. , 22.5, 29. , 24.5, 20. , 18. , 18.5, 17.5, 26.5, 13. , 16.5,
       13. , 13. , 13. , 30. , 25.5, 17.5, 17. , 15

In [111]:
y_mean = np.mean(y_all)
print(y_mean)

23.514572864321607


Найдем $S^2$ — сумму квадратов отклонений наблюдений от общего среднего:

In [112]:
s2 = np.sum((y_all - y_mean)**2)
s2

24252.575477386934

Найдем $S^2_F$ - сумму квадратов отклонений средних групповых значений от общего среднего:

In [113]:
s2_f = ((y1_mean - y_mean)**2) * n1 + ((y2_mean - y_mean)**2) * n2 + ((y3_mean - y_mean)**2) * n3
s2_f

8072.820661929806

Найдем остаточную сумму квадратов отклонений:

In [114]:
s2_residual = np.sum((y1 - y1_mean)**2) + np.sum((y2 - y2_mean)**2) + np.sum((y3 - y3_mean)**2)
s2_residual

16179.754815457129

Удостоверимся, что соблюдается равенство $S^2 = S_F^2 + S_{ост}^2$:

In [115]:
print(s2)
print(s2_f + s2_residual)

24252.575477386934
24252.575477386934


Найдем общую дисперсию:

In [116]:
sigma2_general = s2 / (n - 1)
sigma2_general

61.08961077427439

Найдем факторную дисперсию:

In [117]:
sigma2_f = s2_f / (k - 1)
sigma2_f

4036.410330964903

Найдем остаточную дисперсию:

In [118]:
sigma2_residual = s2_residual / (n - k)
sigma2_residual

40.961404596094

Вычислим $F_H$:

In [119]:
F_h = sigma2_f / sigma2_residual
F_h

98.54179491075868

Найдем значение $F_{крит}$ в таблице критических точек распределения Фишера-Снедекора для заданного уровня значимости $\alpha = 0.05$ и двух степеней свободы:

$df_{межд} = k - 1 = 3 - 1 = 2$ и $df_{внутр} = n - k = 398 - 3 = 395$.

Для данных значений $F_{крит} = 2.9957$. Так как $F_H > F_{крит}$, отвергнута гипотеза Н0, различие среднего роста в трех группах статистически значимо.